File: DATA_STRUCTURES_CPP/Tree_Splay__(Balance_binary_tree).cpp
/*Splay tree - Расширяющееся (косое) самобалансирующееся двоичное дерево поиска*/
#include <iostream>
#include <queue>

using std::queue;
using std::cout;

template <typename Type>
class Splay_tree {
private:
    class Node {
    public:
        Type data;                              // Данные (ключ)
        Node *left, *right, *parent;            // Левый, правый листы и родитель
        Node(const Type& data) :
            data(data), left(nullptr), right(nullptr), parent(nullptr) {};
    };
    Node *root;                                 // Корень
    size_t count;                               // Счетчик элементов
    void  _left_rotate(Node *n);                // Разворот поддерева с корнем n влево
    void  _right_rotate(Node *n);               // Разворот поддерева с корнем n вправо
    void  _splay(Node *n);                      // Поднимет ключ в корень, если он присутствует в дереве, иначе поднимет 
    // в корень самый последний элемент к которому был осуществлен доступ и возвращает новый корень.
    void  _replace(Node* u, Node* v);           // Замена вершин для функции erase, для удаления по ключу
    Node* _subtree_minimum(Node* u);            // Наименьший
    Node* _subtree_maximum(Node* u);            // Наибольший элемент
    Node* _find(const Type &key);               // Поиск по ключу
    void  _print_inord(Node *n); 				// Инфиксный обход
	void  _print_preord(Node *n);				// Префиксный обход
    void  _print_postord(Node *n);				// Постфиксный обход
	void  _delete_bst(Node *n);  				// Рекурсивное удаление всех элементов дерева					
	void  _copy_tree(Node*);					// Итеративное копирование дерева для конструктора копирования
public:
    Splay_tree();                               // Конструктор без параметров
    Splay_tree(const Type &data);				// Конструктор с параметрами
	Splay_tree(const Splay_tree<Type> &tree);	// Конструктор копирования
    ~Splay_tree();                              // Деструктор
    void insert(const Type &key);               // Добавление по ключу
    void erase(const Type &key);                // Удаление по ключу
    void print_inord();							// Инфиксный обход (упорядоченный)  <-- Рекурсивный вариант
	void print_preord();						// Префиксный обход					<-- Рекурсивный вариант
    void print_postord();						// Постфиксный обход	            <-- Рекурсивный вариант
    Type successor(const Type &key);
    Type predecessor(const Type& key);
    const Type& min_key()      { return _subtree_minimum(root)->data; }     // Наименьший элемент
    const Type& max_key()      { return _subtree_maximum(root)->data; }     // Наибольший
    bool empty() const         { return root == nullptr; }                  // Есть элементы или нет
    bool find(const Type &key) { return _find(key) != nullptr; }            // Поиск по значению
    size_t size() const        { return count; }                            // Кол-во элементов
};


template <typename Type>
Splay_tree<Type>::Splay_tree() : root(nullptr), count(0) {};

template <typename Type>
Splay_tree<Type>::Splay_tree(const Type& data) {
    Node *node = new Node(data);
    root = node;
    count = 1;
}

template <typename Type>
Splay_tree<Type>::Splay_tree(const Splay_tree<Type>& tree) {
    _copy_tree(tree.root);
}

template <typename Type>
void Splay_tree<Type>:: _copy_tree(Node* from) {
	queue<Node*> q;
	if (from) {
		q.push(from);
		while (!q.empty()) {
			Node* top = q.front();
			q.pop();
			this->insert(top->data);
			//cout << top->data << " ";
			if (top->left)
				q.push(top->left);
			if (top->right)
				q.push(top->right);
		}
	}
	//cout << std::endl;
}

template <typename Type>
void Splay_tree<Type>::_delete_bst(Node* n) {
    if (n) {
        _delete_bst(n->left);
        _delete_bst(n->right);
        delete n;
    }
}

template <typename Type>
Splay_tree<Type>::~Splay_tree() {
    _delete_bst(root);
}

template <typename Type>
void Splay_tree<Type>::_left_rotate(Node *x) {
    Node* y = x->right;
    if (y) {
        x->right = y->left;
        if (y->left) y->left->parent = x;
        y->parent = x->parent;
    }
    if (!x->parent) 
        root = y;
    else if (x == x->parent->left) 
        x->parent->left = y;
    else 
        x->parent->right = y;
    if (y) 
        y->left = x;
    x->parent = y;
}

template <typename Type>
void Splay_tree<Type>::_right_rotate(Node *x) {
    Node* y = x->left;
    if (y) {
        x->left = y->right;
        if (y->right) y->right->parent = x;
        y->parent = x->parent;
    }
    if (!x->parent) 
        root = y;
    else if (x == x->parent->left)
        x->parent->left = y;
    else 
        x->parent->right = y;
    if (y) 
        y->right = x;
    x->parent = y;
}

template <typename Type>
void Splay_tree<Type>::_splay(Node *x) {
    while (x->parent) {
        if (!x->parent->parent) {
            if (x->parent->left == x) _right_rotate(x->parent);
            else _left_rotate(x->parent);
        }
        else if (x->parent->left == x && x->parent->parent->left == x->parent) {
            _right_rotate(x->parent->parent);
            _right_rotate(x->parent);
        }
        else if (x->parent->right == x && x->parent->parent->right == x->parent) {
            _left_rotate(x->parent->parent);
            _left_rotate(x->parent);
        }
        else if (x->parent->left == x && x->parent->parent->right == x->parent) {
            _right_rotate(x->parent);
            _left_rotate(x->parent);
        }
        else {
            _left_rotate(x->parent);
            _right_rotate(x->parent);
        }
    }
}

template <typename Type>
void Splay_tree<Type>::_replace(Node* u, Node* v) {
    if (!u->parent) root = v;
    else if (u == u->parent->left) u->parent->left = v;
    else u->parent->right = v;
    if (v) v->parent = u->parent;
}

template <typename Type>
typename Splay_tree<Type>::Node* Splay_tree<Type>::_subtree_minimum(Node* n) {
    while (n->left) n = n->left;
    return n;
}

template <typename Type>
typename Splay_tree<Type>::Node* Splay_tree<Type>::_subtree_maximum(Node* n) {
    while (n->right) n = n->right;
    return n;
}

template <typename Type>
typename Splay_tree<Type>::Node* Splay_tree<Type>::_find(const Type &key) {
    Node* z = root;
    while (z) {
        if (z->data < key) z = z->right;
        else if (key < z->data) z = z->left;
        else return z;
    }
    return nullptr;
}

template <typename Type>
void Splay_tree<Type>::insert(const Type& key) {
    Node* z = root;
    Node* p = nullptr;
    while (z) {
        p = z;
        if (z->data < key) z = z->right;
        else z = z->left;
    }
    z = new Node(key);
    z->parent = p;

    if (!p)
        root = z;
    else if (p->data < z->data)
        p->right = z;
    else 
        p->left = z;
    _splay(z);
    count++;
}

template <typename Type>
void Splay_tree<Type>::erase(const Type &key) {
    Node* z = _find(key);
    if (!z) return;

    _splay(z);

    if (!z->left) _replace(z, z->right);
    else if (!z->right) _replace(z, z->left);
    else {
        Node *y = _subtree_minimum(z->right);
        if (y->parent != z) {
            _replace(y, y->right);
            y->right = z->right;
            y->right->parent = y;
        }
        _replace(z, y);
        y->left = z->left;
        y->left->parent = y;
    }
    delete z;
    count--;
}

template <typename Type>
void Splay_tree<Type>::_print_inord(Node *n) {
	if (n) {
		_print_inord(n->left);
		std::cout << n->data << " ";
		_print_inord(n->right);
	}
}

template <typename Type>
void Splay_tree<Type>::print_inord() {
	_print_inord(root);
	std::cout << "\n";
}

template <typename Type>
void Splay_tree<Type>::_print_preord(Node *current) {
	if (current) {
		std::cout << current->data << " ";
		_print_preord(current->left);
		_print_preord(current->right);
	}
}

template <typename Type>
void Splay_tree<Type>::print_preord() {
	_print_preord(root);
	std::cout << "\n";
}

template <typename Type>
void Splay_tree<Type>::_print_postord(Node* current) {
	if (current) {
		_print_preord(current->left);
		_print_preord(current->right);
		std::cout << current->data << " ";
	}
}

template <typename Type>
void Splay_tree<Type>::print_postord() {
	_print_postord(root);
	std::cout << "\n";
}

template <typename Type>
Type Splay_tree<Type>::successor(const Type &key) {
	Node *curr = root;
	Node *succ = nullptr;
	while (curr) {
		if (curr->data > key) {
			succ = curr;
			curr = curr->left;
		}
		else {
			curr = curr->right;
		}
	}
	return (succ) ? succ->data : Type(0);
}

template <typename Type>
Type Splay_tree<Type>::predecessor(const Type& key) {
    Node* curr = root;
    Node* succ = nullptr;
    while (curr) {
        if (curr->data < key) {
            succ = curr;
            curr = curr->right;
        }
        else {
            curr = curr->left;
        }
    }
    return (succ) ? succ->data : Type(0);

}


int main() {
    setlocale(0, "ru");
    Splay_tree<int> stree;
    for (int i = 0; i < 20; ++i)
        stree.insert(i);
    Splay_tree<int> st2(stree);
    cout << "Copy tree:\t";  st2.print_inord();
    cout << "In order:\t";   stree.print_inord();
    cout << "Pre order:\t";  stree.print_preord();
    cout << "Post order:\t"; stree.print_postord();
    cout << "Min elem = " << stree.min_key() << "\n";
    cout << "Max elem = " << stree.max_key() << "\n";
    cout << "Size = "     << stree.size()    << "\n";
    cout << "Find 5 = "   << stree.find(5)   << "\n";
    cout << "Find 100 = " << stree.find(100) << "\n";
    cout << "Empty = "    << stree.empty()   << "\n";
    stree.erase(5);
    stree.erase(15);
    stree.erase(454);
    cout << "delete 5, 15, 454, res == "; stree.print_inord();
    cout << "Succ 10 = "  << stree.successor(10)  << "\n";
    cout << "Succ 100 = " << stree.successor(100) << "\n";
    cout << "Pred 10 = "  << stree.predecessor(10) << "\n";

    auto x = nullptr;
    if (x == nullptr)
        cout << "\nn!!!!\n";
    if (!x)
        cout << "!x И x == NULL - ОДНО И ТОЖЕ\n";

}

File: DATA_STRUCTURES_PY/Tree_Splay__(Balance_binary_tree).py
"""Splay tree - Расширяющееся (косое) самобалансирующиеся двоичное дерево поиска"""

class Node: pass                                # Класс описывающий узлы вставляемые в дерево
class Splay_tree:                               # Расширяющееся самобалансирующееся дерево
    def __init__(self, data=None): pass         # Конструктор с параметром и без
    def _left_rotate(self, x): pass             # Разворот поддерева с корнем x влево
    def _right_rotate(self, x): pass            # Разворот поддерева с корнем x вправо
    def _splay(self, x): pass                   # Поднимет ключ в корень, если он присутствует в дереве, иначе поднимет
    # в корень самый последний элемент к которому был осуществлен доступ и возвращает новый корень.
    def _replace(self, u, v): pass              # Замена вершин для функции erase, для удаления по ключу
    def _subtree_min(self, node): pass          # Наименьший элемент дерева
    def _subtree_max(self, node): pass          # Наибольший элемент
    def _find(self, key): pass                  # Поиск элемента по ключу (Служебная функция)
    def insert(self, key): pass                 # Вставка элемента в дерево
    def erase(self, key): pass                  # Удаление по значению
    def _print_inord(self, node): pass          # Инфиксный обход
    def print_inord(self): pass                 # Печать
    def _print_preord(self, node): pass         # Префиксный обход
    def print_preord(self): pass                # Печать
    def _print_postord(self, node):pass         # Постфиксный обход
    def print_postord(self): pass               # Печать
    def successor(self, key): pass              # Потомок
    def predecessor(self, key): pass            # Предок
    def min_key(self): pass                     # Наименьшее значение в дереве
    def max_key(self): pass                     # Наибольшее
    def find(self, key): pass                   # Поиск по ключу
    def size(self): pass                        # Кол-во элементов
"""-----------------------------------------------------------------------------"""

class Node(object):
    def __init__(self, data):
        self._data = data
        self._left = None
        self._right = None
        self._parent = None


class Splay_tree:
    def __init__(self, data=None):
        if data is None:
            self._root = None
            self._count = 0
        else:
            node = Node(data)
            self._root = node
            self._count = 1

    def _left_rotate(self, x):                # Разворот поддерева с корнем n влево
        y = x._right
        if y:
            x._right = y._left
            if y._left:
                y._left._parent = x
            y._parent = x._parent
        if x._parent is None:
            self._root = y
        elif x == x._parent._left:
            x._parent._left = y
        else:
            x._parent._right = y
        if y:
            y._left = x
        x._parent = y

    def _right_rotate(self, x):
        y = x._left
        if y:
            x._left = y._right
            if y._right:
                y._right._parent = x
            y._parent = x._parent
        if x._parent is None:
            self._root = y
        elif x == x._parent._left:
            x._parent._left = y
        else:
            x._parent._right = y
        if y:
            y._right = x
        x._parent = y

    def _splay(self, x):
        while x._parent:
            if x._parent._parent is None:
                if x._parent._left == x:
                    self._right_rotate(x._parent)
                else:
                    self._left_rotate(x._parent)
            elif x._parent._left == x and x._parent._parent._left == x._parent:
                self._right_rotate(x._parent._parent)
                self._right_rotate(x._parent)
            elif x._parent._right == x and x._parent._parent._right == x._parent:
                self._left_rotate(x._parent._parent)
                self._left_rotate(x._parent)
            elif x._parent._left == x and x._parent._parent._right == x._parent:
                self._right_rotate(x._parent)
                self._left_rotate(x._parent)
            else:
                self._left_rotate(x._parent)
                self._right_rotate(x._parent)

    def _replace(self, u, v):
        if u._parent is None:
            self._root = v
        elif u == u._parent._left:
            u._parent._left = v
        else:
            u._parent._right = v
        if v:
            v._parent = u._parent

    @staticmethod
    def _subtree_min(node):
        while node._left:
            node = node._left
        return node

    @staticmethod
    def _subtree_max(node):
        while node._right:
            node = node._right
        return node

    def _find(self, key):
        curr = self._root
        while curr:
            if curr._data < key:
                curr = curr._right
            elif curr._data > key:
                curr = curr._left
            else:
                return curr
        return None

    def insert(self, key):
        z = self._root
        p = None
        while z:
            p = z
            if key > z._data:
                z = z._right
            else:
                z = z._left
        z = Node(key)
        z._parent = p

        if p is None:
            self._root = z
        elif p._data < z._data:
            p._right = z
        else:
            p._left = z
        self._splay(z)
        self._count += 1

    def erase(self, key):
        z = self._find(key)
        if z is None:
            return
        self._splay(z)
        if z._left is None:
            self._replace(z, z._right)
        elif z._right is None:
            self._replace(z, z._right)
        else:
            y = self._subtree_min(z._right)
            if y._parent != z:
                self._replace(y, y._right)
                y._right = z._right
                y._right._parent = y
            self._replace(z, y)
            y._left = z._left
            y._left._parent = y
        z = None
        self._count -= 1

    def _print_inord(self, node):
        if node:
            self._print_inord(node._left)
            print(node._data, end=' ')
            self._print_inord(node._right)

    def print_inord(self):
        if self._root:
            self._print_inord(self._root)
            print(end='\n')
        else:
            print("BST is empty")

    def _print_preord(self, node):
        if node:
            print(node._data, end=' ')
            self._print_inord(node._left)
            self._print_inord(node._right)

    def print_preord(self):
        if self._count:
            self._print_preord(self._root)
            print(end='\n')
        else:
            print("BST is empty")

    def _print_postord(self, node):
        if node:
            self._print_inord(node._left)
            self._print_inord(node._right)
            print(node._data, end=' ')

    def print_postord(self):
        if self._count:
            self._print_postord(self._root)
            print(end='\n')
        else:
            print("BST is empty")

    def successor(self, key):
        curr = self._root
        succ = None
        while curr:
            if curr._data > key:
                succ = curr
                curr = curr._left
            else:
                curr = curr._right
        return succ._data if succ else None

    def predecessor(self, key):
        curr = self._root
        succ = None
        while curr:
            if curr._data < key:
                succ = curr
                curr = curr._right
            else:
                curr = curr._left
        return succ._data if succ else None

    def min_key(self):
        return self._subtree_min(self._root)._data

    def max_key(self):
        return self._subtree_max(self._root)._data

    def find(self, key):
        return self._find(key) != None

    def size(self):
        return self._count


def main():
    stree = Splay_tree(10)
    for i in range(20):
        stree.insert(i)
    print("In order\t", end='')
    stree.print_inord()
    print("Pre order\t", end='')
    stree.print_preord()
    print("Post order\t", end='')
    stree.print_postord()
    print(f"Max key == {stree.max_key()}")
    print(f"Min key == {stree.min_key()}")
    print(f"Succ 10 == {stree.successor(10)}")
    print(f"Pred 10 == {stree.predecessor(10)}")
    print(f"Size tree: {stree.size()}")


if __name__ == "__main__":
    main()